packages/core/lib/span-context.ts

```diff
@@ -0,0 +1,89 @@
+export interface SpanContext {
+  readonly id: string // 64 bit random string
+  readonly traceId: string // 128 bit random string
+
+  // returns true if this is still considered a valid context
+  readonly isValid: () => boolean
+}
+
+export function spanContextEquals (span1?: SpanContext, span2?: SpanContext) {
+  if (span1 === span2) return true
+
+  if (span1 !== undefined && span2 !== undefined) {
+    return span1.id === span2.id && span1.traceId === span2.traceId
+  }
+
+  return false
+}
+
+export interface SpanContextStorage extends Iterable<SpanContext> {
+  push: (context: SpanContext) => void
+  pop: (context: SpanContext) => void
+  readonly current: SpanContext | undefined
+}
+
+class SpanContextIterator implements Iterator<SpanContext> {
+  private readonly contexts: SpanContext[]
+  private position: number
+
+  constructor (contexts: SpanContext[]) {
+    this.contexts = contexts
+    this.position = this.contexts.length
+  }
+
+  next (): IteratorResult<SpanContext> {
+    if (this.position > 0) {
+      return {
+        done: false,
+        value: this.contexts[this.position--]
+      }
+    } else {
+      return {
+        done: true,
+        value: undefined
+      }
+    }
+  }
+}
+
+export class DefaultSpanContextStorage implements SpanContextStorage {
+  private readonly contextStack: SpanContext[]
+
+  constructor (contextStack: SpanContext[] = []) {
+    this.contextStack = contextStack
+  }
+
+  [Symbol.iterator] (): SpanContextIterator {
+    return new SpanContextIterator(this.contextStack)
+  }
```

Comment:
I believe you can replace the whole `SpanContextIterator` with [a `yield*`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield*):

```suggestion
  *[Symbol.iterator] (): Iterator<SpanContext> {
    return yield* this.contextStack
  }
```

Reply:
Oh nice! TIL!

I've removed that iterator class and switched around the logic for `push`/`pop`/`current` so that we're adding contexts to the start of the array and then the iterator iterates in the correct order

Reply:
Ah I forgot this was in reverse order!

We don't want to have to `shift`/`unshift` the backing array since that's a lot slower than `push`/`pop`, so an implementation like this would be better:

```ts
class DefaultSpanContextStorage {
  *[Symbol.iterator] (): Iterator<SpanContext> {
    for (let i = this.contextStack.length - 1; i >= 0; --i) {
      yield this.contextStack[i]
    }
  }
}
```